substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java

```diff
@@ -1294,6 +1295,11 @@ private static Suites modifySuites(SubstrateBackend backend, Suites suites, Feat
         pos.next();
         pos.add(new StackValuePhase());
 
+        if (hosted) {
+            position = highTier.findPhase(DeadStoreRemovalPhase.class);
+            position.add(new RemoveRedundantClassInitPhase(runtimeCallProviders.getMetaAccess()));
```

Comment:
You have a phase ordering problem here: You insert the phase before inlining because you rely on the ensure initialized method to not be inlined. But that means you are missing a lot of optimization potential because many more checks could be removed after inlining and after high-level optimizations such as loop peeling.

The solution for this is to model class initialization as a separate Graal node that is then lowered to the actual check using a snippet. Then you can insert this phase much lower in the compilation pipeline.

Reply:
Answer related to PR closing. I will move this changelist to private PR, as soon as I get access. That's the reason why I closed this PR in advance, without merge.